envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cve.py

```diff
@@ -0,0 +1,182 @@
+        """CVE nodes used for matching versions."""
+        return self.cve_data['configurations']['nodes']
+
+    @property
+    def published_date(self) -> date:
+        """Date the CVE was published."""
+        return self.parse_cve_date(self.cve_data['publishedDate'])
+
+    @property
+    def score(self) -> float:
+        """CVE score."""
+        return self.cve_data['impact']['baseMetricV3']['cvssV3']['baseScore']
+
+    @property
+    def severity(self) -> str:
+        """CVE severity."""
+        return self.cve_data[
+            'impact']['baseMetricV3']['cvssV3']['baseSeverity']
+
+    @property  # type:ignore
+    @abstracts.interfacemethod
+    def version_matcher_class(self) -> Type[
+            version_matcher.ACVEVersionMatcher]:
+        """Version matcher class."""
+        raise NotImplementedError
+
+    def dependency_match(self, dep: "dependency.ADependency") -> bool:
+        """Heuristically match dependency metadata against CVE.
+
+        In general, we allow false positives but want to keep the noise
+        low, to avoid the toil around having to populate IGNORES_CVES.
+        """
+        # ? wildcard_version_match = False
+        # Consider each CPE attached to the CVE for a match against the
+        # dependency CPE.
+        for cve_cpe in self.cpes:
+            if cve_cpe.dependency_match(dep):
+                # Wildcard version matches need additional heuristics unrelated
+                # to CPE to qualify, e.g. last updated date.
+                return (
+                    self.wildcard_version_match(dep)
+                    if cve_cpe.version == '*'
+                    else True)
+
+    def format_failure(self, dep: "dependency.ADependency") -> str:
+        """Format CVE failure for a given dependency."""
+        return self.fail_template.render(
+            cve=self,
+            dependency=dep)
+
+    def gather_cpes(
+            self,
+            nodes: List["typing.CVENodeDict"],
+            cpe_set: Set["cpe.ACPE"]) -> None:
+        """Recursively gather CPE data from CVE nodes."""
+        for node in nodes:
+            for cpe_match in node.get('cpe_match', []):
+                cpe = self.cpe_class.from_string(cpe_match.pop('cpe23Uri'))
+                if self.include_version(cpe_match, cpe):
+                    cpe_set.add(cpe)
+            children = node.get('children', [])
+            if children:
+                self.gather_cpes(children, cpe_set)
+
+    def include_version(
+            self,
+            cpe_match: "typing.CVENodeMatchDict",
+            cpe: "cpe.ACPE") -> bool:
+        """Determine whether a CPE matches according to installed version of a
+        dependency."""
+        return (
+            str(cpe) in self.tracked_cpes
+            and (
+                self.version_matcher_class(cpe_match)(
+                    self.tracked_cpes[str(cpe)])))
+
+    def parse_cve_date(self, date_str: str) -> date:
+        if not date_str.endswith('Z'):
```

Comment:
@htuch this was one of my concerns

i have switched the `assert` -> `Exception` 

im still thinking that this is not required unless someone is using this to read some non-NIST data that is in NIST format, but for some reason doesnt use UTC times (seems unlikely to me)

Reply:
Yeah, I think assuming NIST actual use for format is fine. If you want to be defensive, leave the assert.